packages/core/nano/src/types.ts

```diff
@@ -1,22 +1,139 @@
+   * Return IDataResult for the given dataBindingReference, containing the data value
+   * and some additional metadata.
+   * @param dataBindingReference - An instance of dataBindingReference, a reference
+   * to some data in the bound component.
+   */
+  read<K extends keyof T>(consumer: IConsumer<T>, property: K, ...args: any[]): T[K] | Pending<T[K]>;
+
+  /**
+   * Invoked when this data provider is deleted.
+   */
+  onDeleted(): void;
+}
+
+/**
+ * The interface for an object that can bind to an IProducer.
+ * One example of this provider-observer relationship is a Formula binding
+ * to Tablero. In this scenario, Formula is an IConsumer listening to changes
+ * in the IProducer, Tablero.
+ *
+ * Any object that implements IConsumer is expected to provide a
+ * callback whenever the component it is bound to changes in value and a reference
+ * to the data that the observer is bound to.
+ */
+export interface IConsumer<T> {
+  /**
+   * Invoked whenever the data this object is bound to is changed.
+   */
+  notify: <K extends keyof T>(property: K, value: T[K]) => void;
+
+  /**
+   * Invoked whenever the provider this is bound to is deleted.
+   */
+  onProducerDeleted: () => void;
 }
 
-export type PickPending<T, K extends keyof T> = { [Key in K]: T[K] | Pending<T[K]> };
-
-export interface Producer<T> {
-    id: string;
-    unsubscribe: (origin: Consumer<T>) => void;
-    
-    now: <K extends keyof T>(property: K, ...args: any[]) => T[K] | undefined;
-    nowMany: <K extends keyof T>(properties: Record<K, undefined | any[]>) => Partial<Pick<T, K>>;
-    
-    request: <K extends keyof T>(origin: Consumer<T>, property: K, ...args: any[]) => T[K] | Pending<T[K]>;
-    requestMany: <K extends keyof T>(origin: Consumer<T>, properties: Record<K, undefined | any[]>) => PickPending<T,K>;
+/**
+ * IDataResult represents the result object passed through data binding between two components.
+ * It provides query functions for retrieving values, hints, friendly names, and provides
+ * basic traverse APIs to move to previous or next level results.
+ */
+export interface IDataResult {
+  /**
+   * Get the list of properties of this dataResult.
+   * @returns a list of dataResults representing the next level objects.
+   */
+  getProperties: (query: string) => Promise<IDataResult[]>;
+
+  /**
+   * Return the raw value of this object.
+   */
+  getValue: () => Pending<IData | undefined>;
+
+  /**
+   * Return the hint text of this object.
+   */
+  getHint: () => string | undefined;
+
+  /**
+   * Return the friendlyName of this object.
+   */
+  getFriendlyName: () => string;
+
+  /**
+   * Return The dataBindingReference for this instance of Producer.
+   */
+  getBindingReference: () => any;
+
+  /**
+   * Get the parent DataResult object.
+   * @returns an instance of IDataResult that represent the previous level (parent) of this object,
+   * or undefined if this is the top level object.
+   */
+  getParent: () => IDataResult | undefined;
```

Comment:
I don't follow 'getParent()'.  Is this related to the parser state at point we evaluated this result?  (Otherwise, wouldn't it be getParents?)

Reply:
Yeah, it's so we can reconstruct the full formula given one piece of it. I'm not sure if this should be more of an implementation detail for the Formula component specifically.